Modify the logic for double-declarative shadow root attachment

mfreed7 opened the pull request:
Per the discussion at https://github.com/whatwg/dom/issues/1235, this PR makes the *first* declarative shadow root "win". Prior behavior was for the *last* one to remain, but this PR changes second and subsequent declarative shadow roots to report an exception instead.

@emilio @annevk @rniwa

- [X] At least two implementers are interested (and none opposed):
   * Chromium (me)
   * WebKit [sounds supportive](https://github.com/whatwg/dom/issues/1235#issuecomment-1900422398)
- [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/shadow-dom/declarative/declarative-shadow-dom-repeats.html
   * https://github.com/web-platform-tests/wpt/pull/44568
- [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://crbug.com/1379513
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1876885
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=268208
- [ ] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: …
- [X] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use.

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10069/parsing.html" title="Last updated on Feb 13, 2024, 9:53 PM UTC (8588e3b)">/parsing.html</a>  ( <a href="https://whatpr.org/html/10069/531e50d...8588e3b/parsing.html" title="Last updated on Feb 13, 2024, 9:53 PM UTC (8588e3b)">diff</a> )

mfreed7 commented on source:
```diff
@@ -126123,21 +126123,17 @@ document.body.appendChild(text);
+     If an exception is thrown, then catch it and <span>report the exception</span>.</p></li>
+
+     <li><p>If an exception was thrown by <span data-x="concept-attach-a-shadow-root">attach a
+     shadow root</span>, or if <var>declarative shadow host element</var> was already a
```
I wasn't totally clear on the best way to structure this, since it's essentially a double-if-otherwise. Let me know how you'd like it to look.

annevk replied:
I think you have two options:

1. You call insert twice.
2. You track success through a variable.

I suspect the first is what we want.

The suggested wording by this PR is quite subtle and it's preferable to have things be very explicit and straightforward to follow.

mfreed7 replied:
Still not exactly sure how you want that structured. I've updated to call insert twice, and added a variable to try to made the double-if structure a little more understandable.

annevk replied:
Hmm, can't you do this:

1. If X is a host: ....
2. Otherwise:
   1. Attach. If an exception is throw, then .... and return.
   2. Let shadow be ...

?

At least looking at this it's not clear to me we need "should attach shadow root".

mfreed7 replied:
Ahh the "and return" bit is the part I was worried about before. But if that's ok, then yes this works. I've updated it accordingly.

annevk reviewed (commented):
Please restore the PR template. This is a normative change and as such we need to go through the PR checkboxes.

mfreed7 commented:
> Please restore the PR template. This is a normative change and as such we need to go through the PR checkboxes.

Done

annevk commented:
@smaug---- @emilio @hsivonen do any of you want to review this as well?

annevk commented:
@mfreed7 I rebased this on top of #10117. Could you have a quick look as to whether it's still good? Also, could you file an MDN issue for the various declarative shadow tree changes?

annevk commented:
I also added/corrected tests here: https://github.com/web-platform-tests/wpt/pull/44568.

mfreed7 commented:
> @mfreed7 I rebased this on top of #10117. Could you have a quick look as to whether it's still good?

Yep, looks good, thanks.

> Also, could you file an MDN issue for the various declarative shadow tree changes?

Sure: https://github.com/mdn/content/issues/32289

> I also added/corrected tests here: https://github.com/web-platform-tests/wpt/pull/44568.

Oh darn, I was working on this one: https://github.com/web-platform-tests/wpt/pull/44562/files

Likely we can combine them. I like your new repeats-2 test. Perhaps we can merge them once one lands.
